### ISSUE TITLE
Is model parallelism supported?

### ISSUE THREAD

**Issue by Dapid** [horovod/horovod#96]
All the examples seem to implement data parallelism. Can I train a model that is too big for my card?

**Comment by alsrgv** [comment]
@Dapid, yes, you can. There are changes you should do:

1. Manually place portions of the model to appropriate GPUs using `with tf.device('/gpu:N'):`

2. Modify or remove the portion of the code that pins GPUs to be visible to TensorFlow: `config.gpu_options.visible_device_list = str(hvd.local_rank())`

3. Modify your `mpirun` command to adjust the number of processes you run per machine. If your model uses all GPUs, you'd run one process per machine.

The rest of the code and documentation should work as is.

**Comment by Dapid** [comment]
Thanks! I have it working now.

This is a rough sketch of how I am defining my model:

```
with tf.device('/gpu:1'):
    inputs = get_inputs()
    for _ in range(n_layers // 2):   
        layer = add_layer(layer)

with tf.device('/gpu:0'):
    for _ in range(n_layers - n_layers // 2):
       layer = add_layer(layer)

model = keras.models.Model(inputs=inputs, outputs=layer)
model.compile()
```
It runs, but `nvidia-smi` reports the usage of GPU0 between 70 and 99%, while GPU1 is between 0 and 4%. Shouldn't the usage be alternating between GPU0 and GPU1? Any insights on this?

Tensorboard tells me that the model is indeed distributed, and it wouldn't fit in a single card anyway.

**Comment by alsrgv** [comment]
@Dapid, looks very reasonable.  Maybe layers on GPU1 are easier to compute, and so the utilization is lower?

**Comment by alsrgv** [comment]
@Nimi42, take a look at this example - https://gist.github.com/alsrgv/cbbb2e25f09c1df983098098511d16c4.  Relevant parts are marked with "Horovod Model Parallelism" comment.

In that example, each process uses two GPUs.  I have 4 GPUs per server, so I launch two processes per server.  I run it with:
```
$ mpirun -np 8 \
    -H server1:2,server2:2,server3:2,server4:2 \
    -bind-to none -map-by slot 
    -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH 
    python train.py
```

The open-source version of OpenMPI directly does not use NCCL 2.  Some of proprietary forks/extensions do.  Horovod uses NCCL 2 independently of OpenMPI.

**Comment by sumedhvdatar** [comment]
Is model parallelism supported in Keras using horovod?

**Comment by alsrgv** [comment]
@sumedhvdatar, yes - use the same technique as in https://gist.github.com/alsrgv/cbbb2e25f09c1df983098098511d16c4 and use https://keras.io/getting-started/faq/#device-parallelism for pinning layers to a specific GPU.

**Comment by sumedhvdatar** [comment]
@alsrgv just made my question more clear by raising this issue #437. Can you clarify on that?

**Comment by brksrl** [comment]
Hello @alsrgv, 

for the given example: https://gist.github.com/alsrgv/cbbb2e25f09c1df983098098511d16c4

As far as I understand this is hybrid parallelism, as there is a model parallelism within each model replica. Hovorod provides MPI communication between replicas, but how about communication between gpu's inside the replica? For the given example above, does "/gpu:0" and "/gpu:1" communicate via MPI between send/receive nodes?

**Comment by brksrl** [comment]
@alsrgv, What if we use cpu instead of gpu. In that case, is communication between "/cpu:0" <-> "/cpu:1" gRPC(default) or MPI? If it is gRPC, is there a way to force communication to be MPI between "/cpu:0" <-> "/cpu:1"?

**Comment by alsrgv** [comment]
@brksrl, if you use CPU, you don't need to manually specify devices.

**Comment by mgoundge11** [comment]
Hi **@alsrgv @Dapid**  I have two systems with single GPU (same gpu's ) each. 
Currently I'm made distributed training running using horovod on these systems using data parallelism approach on my own dataset for object detection using keras retinannet. 
But I wanted to train retinanet using large batch size. Is model parallelism will be helpful? 
If yes, How to do it using horovod? 
**Do I need at-least two GPU's per system to do model parallelism?**
Pls help I'm stuck here. Thank you very much in advance.

**Comment by alsrgv** [comment]
@mgoundge11, Horovod currently supports model parallelism within the server.  It'd be helpful if you can move GPU from one server to another.

**Comment by cofiiwu** [comment]
@alsrgv I am studying model parallelism. In the original tensorflow, even if the colocate_gradients_with_ops of optimizer is enabled, there are some operations of gradient are still located on the default GPU. Can Horovod make sure all operations of FW and BW on the same GPU if I assign the GPU in the model?
[The detail of my question](https://stackoverflow.com/questions/60513983/how-to-place-all-operations-of-gradient-on-the-same-gpu-executes-forwarding)

**Comment by etoilestar** [comment]
> @Nimi42, take a look at this example - https://gist.github.com/alsrgv/cbbb2e25f09c1df983098098511d16c4. Relevant parts are marked with "Horovod Model Parallelism" comment.
> 
> In that example, each process uses two GPUs. I have 4 GPUs per server, so I launch two processes per server. I run it with:
> 
> ```
> $ mpirun -np 8 \
>     -H server1:2,server2:2,server3:2,server4:2 \
>     -bind-to none -map-by slot 
>     -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH 
>     python train.py
> ```
> 
> The open-source version of OpenMPI directly does not use NCCL 2. Some of proprietary forks/extensions do. Horovod uses NCCL 2 independently of OpenMPI.

Hello, please tell me, how did he realize the model parallelism? I don't even find the relevant parameter settings and the way the model is split

**Comment by gg22mm** [comment]
不支持 用进程 来做模型并行 / Process based model parallelism is not supported
